### PR TITLE
Push relabel

### PR DESCRIPTION
**Pull request by kellerjo**
**Changes:**

Adding the **push_relabel** algorithm for maximum flow.
(And later use it to add the **min_s_t_cut** algorithm).

### CODE REVIEW

File: code/graph_algorithms/src/push_relabel/push_relabel.cpp [hunk 1]
```diff
@@ -0,0 +1,191 @@
+#include <vector>
+#include <queue>
+#include <cmath>
+#include <limits>
+
+
+
+// push as much excess flow as possible from f to t
+void push(std::vector<std::vector<int> > * graph, std::vector<std::vector<int> > * flow,
```
Comment (arnavb) [inline code/graph_algorithms/src/push_relabel/push_relabel.cpp]: Don't use pointers, use references if modification is necessary.

Comment (kellerjo) [reply code/graph_algorithms/src/push_relabel/push_relabel.cpp]: thanks for the fast reply, looks like I should have read the codestyle rules before starting.
Made the changes now.

File: code/graph_algorithms/src/push_relabel/push_relabel.cpp [hunk 2]
```diff
@@ -0,0 +1,160 @@
+#include <vector>
```
Comment (arnavb) [inline code/graph_algorithms/src/push_relabel/push_relabel.cpp]: Please follow the coding conventions in our [C++ Style Guide](https://github.com/OpenGenus/cosmos/tree/master/guides/coding_style/c%2B%2B).

File: code/graph_algorithms/src/push_relabel/push_relabel.cpp [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+    flow[f][t] += tmp;
+}
+
+// reset the height of node "node", so excess flow can be pushed
+void relabel(std::vector<std::vector<int> > residual, std::vector<int>& height, int node)
+{
+    int min = std::numeric_limits<int>::max();
+
+    for (int i = 0; i < residual.size(); i++)
+    {
```
Comment (arnavb) [inline code/graph_algorithms/src/push_relabel/push_relabel.cpp]: Remove brackets.

File: code/graph_algorithms/src/push_relabel/push_relabel.cpp [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+
+// expecting propper flow matrix, empty residual and no a ->b, b-> a capacities
+int maxFlowResidual(std::vector<std::vector<int> > graph, std::vector<std::vector<int> >& flow, 
+                      std::vector<std::vector<int> >& residual, int s, int t)
+{
+    int numNodes = graph.size();
+
+    // property check
+    if (graph.size() != graph[0].size() || graph.size() != flow.size())
+        std::cerr << "Graph matrix/Flow Matrix has wrong format" << std::endl;
```
Comment (arnavb) [inline code/graph_algorithms/src/push_relabel/push_relabel.cpp]: Don't use `endl` use `\n`.

File: code/graph_algorithms/src/push_relabel/push_relabel.cpp [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+
+    std::vector<int> excess(numNodes, 0);
+    excess[s] = std::numeric_limits<int>::max();
+
+    // initialize the residual graph
+    for (int i = 0; i < numNodes; i++)
```
Comment (arnavb) [inline code/graph_algorithms/src/push_relabel/push_relabel.cpp]: Use preincrement.

File: code/graph_algorithms/test/push_relabel/test_push_relabel.cpp [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+
+int main()
+{
+	std::cout << "Testing the push-relabel algorithm" << std::endl;
+
+	g_graph.push_back(std::vector<int>{ 0, 1, 0 });
+	g_graph.push_back(std::vector<int>{ 0, 0, 1 });
+	g_graph.push_back(std::vector<int>{ 0, 0, 0 });
+
+	test(1,0,2,true);
```
Comment (arnavb) [inline code/graph_algorithms/test/push_relabel/test_push_relabel.cpp]: Is it possible you could use the built in testing library (Catch): https://github.com/OpenGenus/cosmos/tree/master/test/c%2B%2B?

Comment (kellerjo) [reply code/graph_algorithms/test/push_relabel/test_push_relabel.cpp]: like this?

File: code/graph_algorithms/test/push_relabel/test_push_relabel.cpp [hunk 7]
```diff
@@ -0,0 +1,80 @@
+#define CATCH_CONFIG_MAIN
+#include "../../../../test/c++/catch.hpp"
+#include "../../src/push_relabel/push_relabel.cpp"
+#include <iostream>
+
+
+int g_testCounter = 0;
+std::vector<std::vector<int> > g_graph;
```
Comment (arnavb) [inline code/graph_algorithms/test/push_relabel/test_push_relabel.cpp]: Remove space before `>`.

Comment (kellerjo) [reply code/graph_algorithms/test/push_relabel/test_push_relabel.cpp]: done

**Comment by kellerjo** [comment]
ping

**Comment by arnavb** [comment]
djldk

**Pull request merged**